### PR TITLE
Test and refactor rendering

### PR DESCRIPTION
**Pull request by robbiepitts**
In order to test rendering (which is currently async) we needed to return a promise from `scheduleRerender` and by extension `renderComponent`. What do y'all think about this?

cc @chancancode @tomdale

### CODE REVIEW

File: src/application.ts [hunk 1]
```diff
@@ -51,16 +51,18 @@ export default class Application implements Owner {
...
+  private _scheduled: Option<Promise<void>> = null;
   private _initializers: Initializer[] = [];
   private _initialized = false;
 
   constructor(options: ApplicationOptions) {
     this.rootName = options.rootName;
     this.resolver = options.resolver;
+    this._scheduled = new Promise<void>(resolve => {
+      this._afterRender = resolve;
+    });
```
Comment (robbiepitts) [inline src/application.ts]: This is what allows us to get rid of `_rendered`.

File: test/render-component-test.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+  let containerElement = document.createElement('div');
+
+  let app = buildApp()
+    .template('hello-world', `<h1>Hello Glimmer!</h1>`)
+    .boot();
+
+  return app.renderComponent('hello-world', containerElement).then(() => {
+    assert.equal(containerElement.innerHTML, '<h1>Hello Glimmer!</h1>');
+  });
+});
```
Comment (chancancode) [inline test/render-component-test.ts]: Can you add tests for rendering multiple components, rendering into tags with existing content (with or without nextSibling), etc?

**Comment by robbiepitts** [comment]
@chancancode Just added a bunch of tests. Let me know how this looks.

**Comment by tomdale** [comment]
`scheduleRerender` is a potentially very hot path, so we need to be careful not to do anything to deopt it.

I'm OK with this returning a promise, but:

1. Should we have an optimized path that doesn't allocate a `Promise`? The reality is that this is getting called potentially every rAF and the internals don't require the promise. We should avoid the allocation if we don't actually need it.
2. We should continue to maintain a Boolean flag, and make the check more explicit, c.f. https://twitter.com/michaelhaeu/status/845003383153025024 and https://github.com/developit/preact/pull/610.

**Comment by chancancode** [comment]
@robbiepitts @tomdale 

Re: (2) the micro-optimization from Benedikt's slide isn't actually referring to this case (and it likely won't matter here). If we care about that, the right thing to do here is to check for `(this._scheduled !== null)`. Maintaining an parallel boolean field would likely turn out to be more costly (again, I doubt that matters at all).

We can do (1) though. Something like this would probably work:

```typescript
function NOOP() {}

class Application {
  private rerender: () => void = NOOP;
  private afterRender: () => void = NOOP;
  private renderPromise: Option<Promise<void>> = null;

  constructor() {
    // ...
  }

  render() {
    // ...
    this.env.commit();

    let renderResult = result.value;

    this.rerender = () => {
      this.env.begin();
      renderResult.rerender();
      this.env.commit();
      this.didRender();
    };

    this.didRender();
  }

  didRender(): void {
    let { afterRender } = this;

    this.afterRender = NOOP;
    this.renderPromise = null;

    afterRender();
  }

  scheduleRerender(): Promise<void> {
    let { renderPromise } = this;

    if (renderPromise === null) {
      renderPromise = this.renderPromise = new Promise<void>(resolve =>{
        this.afterRender = resolve;
      });

      this._scheduleRerender();
    }

    return renderPromise;
  }

  // internal version used by @tracked, etc
  _scheduleRerender() {
    if (this.renderPromise !== null) {
      requestAnimationFrame(this.rerender);
    }
  }
}
```

Does that make sense?

**Comment by robbiepitts** [comment]
Just updated this per @chancancode's revision. I feel like I'm missing something though. I'm not seeing a path that avoids the `Promise` allocation. You can call `_scheduleRerender` _once_ after the initial render, but then after that it's a noop.

**Comment by robbiepitts** [comment]
My latest commit adds the promise-free path along with a couple boolean flags that seemed necessary to me. Essentially, `_scheduleRerender` has turned into the old `scheduleRerender` and the new `scheduleRerender` wraps it in a promise.

**Comment by chancancode** [comment]
Looks good to me! @krisselden might be able to comment on what patterns are good/bad performance-wise, but that could happen async and we don't need to block on that. @tomdale seems good to you?

**Comment by tomdale** [comment]
This looks good. My one request is that sooner rather than later we think generally about how we want to expose global rendering/lifecycle events. There are a number of things that are currently internal that people might want for testing, analytics, performance measuring, etc. I'd prefer to think about the problem holistically instead of accreting promise-based APIs for every use case that crops up.

**Pull request merged**